File: java/53maxSubArray/53.最大子序和.java
/*
 * @lc app=leetcode.cn id=53 lang=java
 *
 * [53] 最大子序和
 */

// @lc code=start
class Solution {
    public int maxSubArray(int[] nums) {
        int maxSoFar = nums[0];
        int maxEndingPoint = nums[0];
        for (int i = 1; i < nums.length; i++) {
            maxEndingPoint = Math.max(maxEndingPoint + nums[i], nums[i]);
            maxSoFar = Math.max(maxSoFar, maxEndingPoint);
        }
        return maxSoFar;
    }
}
// @lc code=end

File: go/53maxSubArray/53.最大子序和.go
/*
 * @lc app=leetcode.cn id=53 lang=golang
 *
 * [53] 最大子序和
 */

// @lc code=start
func maxSubArray(nums []int) int {
	len := len(nums);
	if len == 0 {
		return 0;
	}
	max, temp := nums[0], nums[0]
	for i := 1; i < len; i++ {
		temp =  maxInt(temp, 0) + nums[i]
		if temp > max {
			max = temp
		}
	}
	return max
}

func maxInt(a int, b int) int {
	if a >= b {
		return a
	} else {
		return b
	}
}
// @lc code=end

